### PR TITLE
Improve push update options

### PR DESCRIPTION
**Pull request by lunny**
As title.

### CODE REVIEW

**Review by guillep2k — APPROVED** [review]

**Comment by guillep2k** [comment]
Tests require updating.

**Review by zeripath — COMMENTED** [review]
I would guess that the functions should also be exported rather than private

**Review by adelowo — APPROVED** [review]

File: modules/repofiles/update.go [hunk 1]
```diff
@@ -455,12 +455,58 @@ type PushUpdateOptions struct {
 	Branch       string
 }
 
+func (opts PushUpdateOptions) isNewRef() bool {
```
Comment (6543) [inline modules/repofiles/update.go]: can we move https://github.com/go-gitea/gitea/pull/10105/files#diff-82f7da3d2fe2bbdd9efe0e6103278d26R458-R505 to a new file. perhaps modules/repofiles/checks.go or so?

**Comment by lunny** [comment]
@zeripath I have exported all the functions. @6543 maybe another package `structs`? But I think it's not this PR.

**Review by zeripath — APPROVED** [review]

File: modules/repofiles/update.go [hunk 2]
```diff
@@ -452,15 +452,77 @@ type PushUpdateOptions struct {
...
+	return opts.IsBranch() && opts.IsNewRef()
+}
+
+// IsUpdateBranch return true if it's not the first push to a branch
+func (opts PushUpdateOptions) IsUpdateBranch() bool {
+	return opts.IsBranch() && opts.IsUpdateRef()
+}
+
+// IsDelBranch return true if it's a deletion to a branch
+func (opts PushUpdateOptions) isDelBranch() bool {
```
Comment (zeripath) [inline modules/repofiles/update.go]:
```suggestion
func (opts PushUpdateOptions) IsDelBranch() bool {
```

**Comment by 6543** [comment]
> @zeripath I have exported all the functions. @6543 maybe another package `structs`? But I think it's not this PR.

Hmm OK

**Review by lafriks — APPROVED** [review]

File: modules/repofiles/update.go [hunk 3]
```diff
@@ -576,26 +686,24 @@ func createCommitRepoActions(repo *models.Repository, gitRepo *git.Repository, o
...
 	for _, opts := range optsList {
-		isNewRef := opts.OldCommitID == git.EmptySHA
-		isDelRef := opts.NewCommitID == git.EmptySHA
-		if isNewRef && isDelRef {
+		if opts.IsNewRef() && opts.IsDelRef() {
 			return nil, fmt.Errorf("Old and new revisions are both %s", git.EmptySHA)
 		}
 		var commits = &repo_module.PushCommits{}
-		if strings.HasPrefix(opts.RefFullName, git.TagPrefix) {
+		if opts.IsNewTag() {
```
Comment (lunny) [inline modules/repofiles/update.go:693]: This introduced a bug. It should be `opts.IsTag()`.  #10663 will resolve the problem.

**Pull request merged**